src/io/io_hdf5.nim

```diff
@@ -0,0 +1,191 @@
+  let convertTo = h5dset.convertType(T)
+  # finally convert seq to tensor and reshape
+
+  # Read the data
+  tensorCpu(shape, result, layout)
+  result.data = convertTo(h5dset)
+
+  assert shape == h5dset.shape
+  assert shape == @(result.shape)
+  assert rank == result.rank
+  assert size == result.size
+
+proc read_hdf5*[T: SomeNumber](h5f: var H5FileObj,
+                              name, group = "",
+                              number = -1): Tensor[T] {.noInit, inline.} =
+  ## wrapper around the real `read_hdf5` to provide a nicer interface
+  ## without having to worry about `some` and `none`
+  let
+    nameOpt = if name.len > 0: some(name) else: none(string)
+    groupOpt = if group.len > 0: some(group) else: none(string)
+    numberOpt = if number >= 0: some(number) else: none(int)
+  result = read_hdf5[T](h5f, nameOpt, groupOpt, numberOpt)
+
+proc read_hdf5*[T: SomeNumber](hdf5Path: string,
+                               name, group = "",
+                               number = -1): Tensor[T] {.noInit, inline.} =
+  ## convenience wrapper around `read_hdf5` with `var H5DataSet` argument.
+  ## opens the given H5 file for reading and then calls the read proc
+  withH5(hdf5Path, "r"):
+    # template opens file with read access as injected `h5f` and closes
+    # file after actions
+    result = read_hdf5[T](h5f, name, group, number)
+
+proc write_hdf5*[T: SomeNumber](h5f: var H5FileObj,
+                                t: Tensor[T],
+                                name, group: Option[string]) =
+  ## Exports a tensor to a hdf5 file
+  ## To keep this a simple convenience proc, the tensor is stored
+  ## in the root group of the hdf5 file under a generic name.
+  ## If the `name` argument is given, the tensor is stored under this
+  ## name instead. If the `group` argument is given the tensor is
+  ## stored in that group instead of the root group.
+  ## Note: if no `name` is given, we need to visit the whole file
+  ## to check for existing tensors. This will introduce a small
+  ## overhead!
+  ##
+  ## Input:
+  ##   - The tensor to write
+  ##   - The H5 file object we write to
+  ##   - An optional name for the dataset of the tensor in the file
+  ##     Useful to store multiple tensors in a single HDF5 file
+  ##   - An optional name for a `group` to store the tensor in
+
+  let (dsetName, grpName, numTensors) = h5f.parseNameAndGroup(true, name, group)
+
+  var dset = h5f.create_dataset(grpName / dsetName,
+                                @(t.shape),
+                                dtype = T)
+
+  # make sure the tensor is contiguous
+  let tCont = t.asContiguous
+  # write tensor data
+  dset.unsafeWrite(tCont.get_data_ptr, tCont.size)
+
+  # now write attributes of tensor
+  dset.attrs["rank"] = tCont.rank
+  dset.attrs["shape"] = @(tCont.shape)
+  dset.attrs["size"] = tCont.size
+  # workaround since we can't write bool attributes
+  dset.attrs["is_C_contiguous"] = if tCont.is_C_contiguous: "true" else: "false"
```

Comment:
In which case is this needed? It's forced contiguous on line 159?

Reply:
Oh, good point. I'm not entirely sure anymore what I had in mind. I guess to store whether the original tensor was contiguous or not.